### PR TITLE
directly include the sildisco module

### PR DESCRIPTION
**Pull request by briskt**
### Added
- Copied all pertinent code, configuration, and tests from [silinternational/simplesamlphp-module-sildisco](https://github.com/silinternational/simplesamlphp-module-sildisco).
- Added new containers `ssp-idp3.local`, `ssp-sp2.local`, `ssp-sp3.local`, `dynamo`, and `init-dynamo` for test and development.

### Changed
- Changed the auth source on idp2 to exampleauth. The sildisco tests originally used admin but that conflicted with other tests that required the use of exampleauth.

### Removed
- Remove test code that was duplicated across multiple modules' tests.

_Note: the [first](https://github.com/silinternational/ssp-base/commit/43ad200cf37947e5420eec222419c33c95dd74b9) commit is mostly just copied files, and the remainder are the changes made to those files._

### CODE REVIEW

File: features/expirychecker.feature [hunk 1]
```diff
@@ -5,29 +5,29 @@ Feature: Expiry Checker module
 
   Scenario: Password will expire in the distant future
     Given I provide credentials that will expire in the distant future
-    When I login
+    When I log in
```
Comment (forevermatt) [inline features/expirychecker.feature:8]: 😄 Nice

File: development/idp-local/config/authsources.php [hunk 2]
```diff
@@ -1186,5 +1186,29 @@
             ],
             'manager_email' => ['manager@example.com'],
         ],
+
+        // sildisco test user
+        'sildisco_idp1:sildisco_password' => [
+            'eduPersonPrincipalName' => ['sildisco@idp1'],
+            'eduPersonTargetID' => ['57de1930-c5d2-4f6f-9318-d85a939c45d8'],
```
[the remark this answers is not in the document]
Comment (briskt) [reply development/idp-local/config/authsources.php:1193]: I see we have an issue in YT for this, so I won't make any changes in this PR. Thanks for mentioning it to help keep it in mind, though.

File: development/sp3-local/config/authsources.php [hunk 3]
```diff
@@ -27,9 +27,12 @@
...
         // Can be NULL/unset, in which case a builtin discovery service will be used.
         'discoURL' => null,
-    ),
 
-    'hub4tests' => array(
+        // Specify what private key to use (such as for decrypting assertions).
+        'privatekey' => 'ssp-hub-sp3.pem',
+    ],
+
+    'ssp-hub-custom-port' => [
```
[the remark this answers is not in the document]
Comment (briskt) [reply development/sp3-local/config/authsources.php:35]: Probably not, thanks for identifying it.

File: development/sp3-local/config/config.php [hunk 4]
```diff
@@ -85,11 +120,11 @@
     'errorreporting' => false,
 
     /*
-     * Custom error show function called from \SimpleSAML\Error\Error::show.
+     * Custom error show function called from SimpleSAML_Error_Error::show.
```
[the remark this answers is not in the document]
Comment (briskt) [reply development/sp3-local/config/config.php:123]: I copied the content of sp2-local/config/config.php into this one, thinking consistency was better than whatever might have been in the sildisco repo. Seeing this now, I see that I overlooked some details. But since it looks like this type of error is only in comments, I think I'll leave it. We have another ticket to make sure files like this are updated using new SSP templates.

File: docker-compose.yml [hunk 5]
```diff
@@ -334,7 +342,7 @@ services:
       # Utilize custom metadata
       - ./development/sp3-local/metadata/saml20-idp-remote.php:/data/vendor/simplesamlphp/simplesamlphp/metadata/saml20-idp-remote.php
     ports:
-      - "8083:80"
+      - "8084:80"
```
[the remark this answers is not in the document]
Comment (briskt) [reply docker-compose.yml:345]: I agree. I'd been holding off on that for later optimization. Maybe now's the time, though. Feeling better about it now that I'm getting a fresh start for the day.

File: docs/development.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
+Three SPs, a hub (a combined IdP and SP) and three IdPs get spun up by docker-compose.  In order for this to work, you will need to edit your hosts file to include entries for the following domains ...
+* ssp-sp1.local   # to be used with port 8081
+* ssp-sp2.local  # to be used with port 8082
+* ssp-sp3.local  # to be used with port 8084
```
[the remark this answers is not in the document]
Comment (briskt) [reply docs/development.md:4]: Same as above content. I think I'll go ahead and clean this up now.

File: docs/overview.md [hunk 7]
```diff
@@ -0,0 +1,4 @@
+The sildisco module is a module for simplesamlphp.  It's main purpose is to allow a simplesamlphp hub to control which Service Providers (SP's) can see and authenticate through which Identity Providers (IdP's).
+
+It relies on some utilities found in [ssp-utilities](https://github.com/silinternational/ssp-utilities).
```
[the remark this answers is not in the document]
Comment (briskt) [reply docs/overview.md:3]: Hmm, I'll take a look. Might want to move that whole repo into this one as well.

File: docs/the_hub.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
+The hub will need its certs, `config.php` and `authsources.php` files as a normal simplesamlphp installation. Examples of these can be found in the `./development/hub` folder. (Note the `discoURL` entry in the `authsources.php` file.)
+
+Other files it will need are as follows ...
+* `/data/vendor/simplesamlphp/simplesamlphp/modules/sildisco/default-enable` needs to be created (just an empty file)
```
[the remark this answers is not in the document]
Comment (briskt) [reply docs/the_hub.md:4]: Good catch

[the remark this answers is not in the document]
Comment (briskt) [reply docs/the_hub.md:4]: But it doesn't become required until SSP 2.0

File: docs/the_hub.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+* The files in the `./www` folder will need to go into `/data/vendor/simplesamlphp/simplesamlphp/modules/sildisco/www`
+* The `./sspoverrides/www_saml2_idp/SSOService.php` file will need overwrite the same out-of-the-box file in `/data/vendor/simplesamlphp/simplesamlphp/www/saml2/idp/`
+
+### Metadata files
+The hub should use the `saml20-*-remote.php`  files from [ssp-base](https://github.com/silinternational/ssp-base) in `/data/vendor/simplesamlphp/simplesamlphp/metadata/`.  These pull in metadata from all the files named `idp-*.php` and `sp-*.php` respectively, including those in sub-folders.
+
+In order for forced re-authentication to be limited only to situations which warrant it, the `saml20-idp-hosted.php` file should include an authproc as such ...
+>  [
+>     'class' =>'sildisco:TrackIdps',
+>  ]
```
[the remark this answers is not in the document]
Comment (briskt) [reply docs/the_hub.md:15]: I see this comment at the top of the TrackIdps.php file:

>  * Attribute filter for adding Idps to the session

I think you're thinking of the LogUser authproc.

File: docs/development.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (forevermatt) [reply docs/development.md]: Sounds good. Thanks for doing that. 👍

File: docs/the_hub.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (forevermatt) [reply docs/the_hub.md:14]: Ahh, okay. Thanks.

File: Dockerfile [hunk 10]
```diff
@@ -56,7 +56,7 @@ COPY dockerbuild/ssp-overrides/id.php $SSP_PATH/www/id.php
 COPY dockerbuild/ssp-overrides/announcement.php $SSP_PATH/announcement/announcement.php
 COPY tests /data/tests
 
-RUN cp $SSP_PATH/modules/sildisco/sspoverrides/www_saml2_idp/SSOService.php $SSP_PATH/www/saml2/idp/
+RUN cp $SSP_PATH/modules/sildisco/lib/SSOService.php $SSP_PATH/www/saml2/idp/
```
Comment (forevermatt) [inline Dockerfile:59]: This is more for my own education, but why the change from sspoverrides to lib?

**Review by forevermatt — APPROVED** [review]

File: Dockerfile (hunk 10 above)
Comment (briskt) [reply Dockerfile:59]: Unfortunately, I don't remember if it was a requirement or a preference. My understanding as of today is that it's not a good idea to define our own directory name(s) inside a module directory structure. `lib` is defined in SimpleSAMLphp documentation, whereas `sspoverrides` is not. One alternative is to put it in an `sspoverrides` directory (or the existing `dockerbuild/ssp-overrides` at the top level of the repository, but that disconnects it from being associated with the sildisco module. I'm open to suggestions, but I'll go ahead and merge this and implement any changes in a later PR.

Comment (forevermatt) [reply Dockerfile:59]: Okay, that makes sense. As I was reading this PR, it was a little hard to tell where we were copying that file _from_ vs. _to_, and having "sspoverrides" in our _from_ path made that detail more obvious. That's all.
